crates/energy/src/graph.rs

```diff
@@ -0,0 +1,242 @@
+                Point::from(*(transform.translation + Vec3::splat(MAX_DISTANCE)).as_ref()),
+            );
+
+            let producers = spacial_index_producer.query_aabb(aabb, Some(entity));
+
+            let receivers = spacial_index_receiver.query_aabb(aabb, Some(entity));
+
+            update_nearby(nearby_units, producers.collect(), receivers.collect());
+        });
+    println!("update_nearby_recv: {:?}", time.elapsed());
+}
+
+fn update_nearby(
+    mut nearby_units: Mut<NearbyUnits>,
+    producers: Vec<Entity>,
+    receivers: Vec<Entity>,
+) {
+    let mut nearby_producers = SmallVec::new();
+    let mut nearby_receivers = SmallVec::new();
+
+    for producer in producers {
+        nearby_producers.push(producer);
+    }
+
+    for receiver in receivers {
+        nearby_receivers.push(receiver);
+    }
+
+    nearby_units.0 = smallvec![
+        Nearby::Producer(nearby_producers),
+        Nearby::Receiver(nearby_receivers)
+    ];
+}
+
+#[allow(clippy::type_complexity)]
+fn update_graph(
+    mut power_grid: ResMut<PowerGrid>,
+    nearby_units: Query<(Entity, &NearbyUnits), Or<(Added<NearbyUnits>, Changed<NearbyUnits>)>>,
+) {
+    for (entity, nearby_units) in nearby_units.iter() {
+        let mut edges_to_remove = HashSet::new();
+
+        edges_to_remove.extend(
+            &mut power_grid
+                .graph
+                .edges(entity)
+                .map(|edge| (edge.source(), edge.target())),
+        );
+
+        let mut edges_to_add = vec![];
+        for group in &nearby_units.0 {
+            match group {
+                Nearby::Producer(producers) => {
+                    for producer in producers {
+                        edges_to_add.push((entity, *producer));
+                    }
+                }
+                Nearby::Receiver(receivers) => {
+                    for receiver in receivers {
+                        edges_to_add.push((entity, *receiver));
+                    }
+                }
+            }
+        }
+
+        for edge in edges_to_add.iter() {
+            if !power_grid.graph.contains_edge(edge.0, edge.1) {
+                power_grid.graph.add_edge(edge.0, edge.1, 1.0);
+            }
+            if edges_to_remove.contains(edge) {
+                edges_to_remove.remove(edge);
+            }
+        }
+
+        for edge in edges_to_remove.iter() {
+            power_grid.graph.remove_edge(edge.0, edge.1);
+        }
+    }
+}
+
+fn remove_old_nodes(
+    mut power_grid: ResMut<PowerGrid>,
+    mut nearby_units: Query<&mut NearbyUnits>,
+    mut death_events: EventReader<DespawnedComponentsEvent<NearbyUnits>>,
+) {
+    for event in death_events.iter() {
+        power_grid.graph.remove_node(event.entity);
+
+        for outer_nearby in event.data.0.iter().flat_map(|nearby| nearby.clone().into_inner()) {
+            for inner_nearby in nearby_units.get_mut(outer_nearby).unwrap().0.iter_mut(){
```

Comment:
This seems a bit convoluted. I would create methods in `Nearby` for removal of entities so that we do not have to call `retain()` in this system. Furthermore, `event.data.0` can have `producers` and `receivers`, these can be iterated so that opposite `producer -> receiver` can be updated in the other entity.

Reply:
I understand our first point but I'm not sure what you mean for your second point.

Reply:
The second point is that while iterating over `producers`, you only need to update `receivers` of pointed to entities. And vice versa.